components/raftstore-v2/src/operation/command/admin/split.rs

```diff
@@ -0,0 +1,621 @@
+        &mut self,
+        req: &AdminRequest,
+        log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        PEER_ADMIN_CMD_COUNTER.batch_split.all.inc();
+
+        let region = self.region_state().get_region();
+        let region_id = region.get_id();
+        validate_batch_split(req, self.region_state().get_region())?;
+
+        let mut boundaries: Vec<&[u8]> = Vec::default();
+        boundaries.push(self.region_state().get_region().get_start_key());
+        for req in req.get_splits().get_requests() {
+            boundaries.push(req.get_split_key());
+        }
+        boundaries.push(self.region_state().get_region().get_end_key());
+
+        info!(
+            self.logger,
+            "split region";
+            "region" => ?region,
+            "boundaries" => %KeysInfoFormatter(boundaries.iter()),
+        );
+
+        let split_reqs = req.get_splits();
+        let new_region_cnt = split_reqs.get_requests().len();
+        let new_version = region.get_region_epoch().get_version() + new_region_cnt as u64;
+
+        let mut derived_req = SplitRequest::default();
+        derived_req.new_region_id = region.id;
+        let derived_req = &[derived_req];
+
+        let right_derive = split_reqs.get_right_derive();
+        let reqs = if right_derive {
+            split_reqs.get_requests().iter().chain(derived_req)
+        } else {
+            derived_req.iter().chain(split_reqs.get_requests())
+        };
+
+        let regions: Vec<_> = boundaries
+            .array_windows::<2>()
+            .zip(reqs)
+            .map(|([start_key, end_key], req)| {
+                let mut new_region = Region::default();
+                new_region.set_id(req.get_new_region_id());
+                new_region.set_region_epoch(region.get_region_epoch().to_owned());
+                new_region.mut_region_epoch().set_version(new_version);
+                new_region.set_start_key(start_key.to_vec());
+                new_region.set_end_key(end_key.to_vec());
+                new_region.set_peers(region.get_peers().to_vec().into());
+                // If the `req` is the `derived_req`, the peers are already set correctly and
+                // the following loop will not be executed due to the empty `new_peer_ids` in
+                // the `derived_req`
+                for (peer, peer_id) in new_region
+                    .mut_peers()
+                    .iter_mut()
+                    .zip(req.get_new_peer_ids())
+                {
+                    peer.set_id(*peer_id);
+                }
+                new_region
+            })
+            .collect();
+
+        let derived_index = if right_derive { regions.len() - 1 } else { 0 };
+
+        // We will create checkpoint of the current tablet for both derived region and
+        // split regions. Before the creation, we should flush the writes and remove the
+        // write batch
+        self.flush();
+        self.write_batch_mut().take();
```

Comment:
Why is this necessary?

Reply:
The wb now may be belong to the old tablet, but we have new tablet now. If it is not taken, then the future writes are still written in the old tablet.

Reply:
Should add a new comment. Because the reasoning is different from "persist all data to disk".

Reply:
BTW, why need to replace the leftmost tablet at all? Couldn't we reuse the source tablet?

Reply:
> BTW, why need to replace the leftmost tablet at all? Couldn't we reuse the source tablet?

tablet suffix will be changed.

Reply:
> Should add a new comment. Because the reasoning is different from "persist all data to disk".

Ok, done.

Reply:
That's not the point, technically you can still use the same tablet and insert it as a new suffix. For reads on the old tablet, we can make sure its snapshot timestamp never exceeds the current maximum sequence number.

Reply:
> That's not the point, technically you can still use the same tablet and insert it as a new suffix. For reads on the old tablet, we can make sure its snapshot timestamp never exceeds the current maximum sequence number.

Split here is not done. If load it with the new suffix, and the split failed, there may be some problem. In this implementation, if that case happens, just remove the these tablets.

Reply:
Fair